src/core/qgsfeatureiterator.cpp

```diff
@@ -121,35 +121,37 @@ void QgsAbstractFeatureIterator::geometryToDestinationCrs( QgsFeature &feature,
   }
 }
 
-void QgsAbstractFeatureIterator::updateRequestToSourceCrs( QgsFeatureRequest &request, const QgsCoordinateTransform &transform ) const
+QgsAbstractFeatureIterator::RequestToSourceCrsResult QgsAbstractFeatureIterator::updateRequestToSourceCrs( QgsFeatureRequest &request, const QgsCoordinateTransform &transform ) const
 {
+  if ( transform.isShortCircuited() )
+    return RequestToSourceCrsResult::Success; // nothing to do
+
   switch ( request.spatialFilterType() )
   {
     case Qgis::SpatialFilterType::NoFilter:
-      break;
+      return RequestToSourceCrsResult::Success;
+
     case Qgis::SpatialFilterType::BoundingBox:
     {
       QgsRectangle newRect = transform.transformBoundingBox( request.filterRect(), Qgis::TransformDirection::Reverse );
       request.setFilterRect( newRect );
-      break;
+      return RequestToSourceCrsResult::Success;
     }
     case Qgis::SpatialFilterType::DistanceWithin:
     {
-      QgsGeometry geom = request.referenceGeometry();
-      QgsRectangle bbox = geom.boundingBox();
-      double x0 = bbox.xMaximum();
-      double y0 = bbox.yMaximum();
-      double dist = request.distanceWithin();
-      QgsPoint p0( x0, y0 );
-      QgsPoint p1( x0 + dist, y0 );
-      QgsLineString newDistLine( p0, p1 );
-      newDistLine.transform( transform, Qgis::TransformDirection::Reverse );
-      dist = newDistLine.length();
-      geom.transform( transform, Qgis::TransformDirection::Reverse );
-      request.setDistanceWithin( geom, dist );
-      break;
+      // we can't safely handle a distance within query, as we cannot transform the
+      // static within tolerance distance from one CRS to a static distance in a different CRS.
```

Comment:
I do understand my implementation was an approximation of the actual distance conversion, but I'd love to see a test failing when using that approximation. It would be useful to have such test in place at least showing why my implementation was not acceptable.